src/react-web3/Web3Provider.js

```diff
@@ -12,72 +12,95 @@ class Web3Provider extends Component {
     this.state = {
       selectedAccount: null,
       web3: null,
-      render: false
+      approvePermissions: null,
+      render: null
     }
 
-    this.interval = null
   }
 
   render() {
     const { web3UnavailableScreen: Web3UnavailableScreen } = this.props
-    let renderContainer = false
 
-    if (this.state.render) {
-      if (this.state.web3) {
-        renderContainer = this.props.children
-      } else {
-        renderContainer =  <Web3UnavailableScreen/>
-      }
-    } else {
-      renderContainer = <Loader show={true}></Loader>
+    if (this.state.render && this.state.web3 && this.state.approvePermissions) {
+      return this.props.children
+    }
+
+    if (!this.state.render && (!this.state.web3 || !this.state.approvePermissions)) {
+      return <Loader show={true} />
     }
-    return ( renderContainer )
 
+    if (this.state.render && (!this.state.web3 || !this.state.approvePermissions)) {
+      return <Web3UnavailableScreen />
+    }
+
+    return null
   }
 
   componentDidMount() {
     this.checkWeb3()
-    this.initPoll()
-
-    // Wait 1 second to render
-    setTimeout(function() {
-      this.setState({render: true})
-    }.bind(this), 1500)
+      .then(this.initPoll())
+      .then(() => {
+        this.setState({
+          render: true
+        })
+      })
+      .catch(err => {
+        console.log(err)
+        this.setState({
+          render: true
+        })
+      })
   }
 
   checkWeb3() {
     const setWeb3 = () => {
-      window.web3 = new Web3(window.web3.currentProvider)
-      this.setState({
-        web3: window.web3
-      })
-      this.fetchAccounts()
+      try {
+        window.web3 = new Web3(window.web3.currentProvider)
+        this.setState({
+          approvePermissions: true,
+          web3: window.web3
+        })
+      } catch (err) {
+        console.log('There was a problem fetching accounts', err)
+      }
     }
 
-    window.addEventListener('load', () => {
-      const { ethereum } = window
-      // Modern dapp browsers...
-      if (ethereum) {
-        window.web3 = new Web3(ethereum)
-        try {
-          // Request account access if needed
-          ethereum.enable().then(()=> {
-            if (!this.state.web3) {
-              setWeb3()
-            }
-          })
-        } catch (error) {
-          // User denied account access...
-          console.log('User denied account access')
+    return new Promise((resolve, reject) => {
```

Comment:
Why a `Promise` with `throw` and `reject`/`resolve` mixed?

Can it be done entirely with `async/await` or `resolve/reject`?